README.md
Sentry in Docker
==================

## Warning about build tags

Latest changes introduced some new build tags:

* **8.0**  - current stable version (8.0.X)
* **7.7**  - old stable version (7.7.4) - no longer updated
* **7.6**  - old stable version (7.6.2) - no longer updated
* **7.5**  - old stable version (7.5.6) - no longer updated
* **7.4**  - old stable version (7.4.3) - no longer updated
* **7.3**  - old stable version (7.3.2) - no longer updated
* **7.2**  - old stable version (7.2.0) - no longer updated
* **7.1**  - old stable version (7.1.4) - no longer updated
* **7.0**  - older stable version (7.0.2) - no longer updated
* **6.4** - even older stable version (6.4.4) - no longer updated
* **dev** - current master on github (infrequent builds)
* **latest** (the default one used earlier) - is now the same as **8.0**

if you want to keep your builds same as before update your Dockerfiles and change
```FROM slafs/sentry``` to ```FROM slafs/sentry:6.4```.

[![Requirements Status](https://requires.io/github/slafs/sentry-docker/requirements.png?branch=master)](https://requires.io/github/slafs/sentry-docker/requirements/?branch=master)

This is my approach for running [Sentry](https://getsentry.com) inside [Docker](https://docker.com/).
Almost everything here is configurable via environment variables (including DATABASES and CACHES settings).
It can be easily configured to run with redis (cache, buffers and celery broker), postgres database, LDAP and REMOTE_USER authentication backends.

## Quickstart

to run a Sentry instance with default settings (with sqlite, locmem cache and no celery) run:

```
docker run -d --name=sentry --volume=/tmp/sentry:/data -p 80:9000 -e SECRET_KEY=randomvalue -e SENTRY_URL_PREFIX=http://sentry.mydomain.com slafs/sentry
```

You can visit now http://sentry.mydomain.com (assuming ``sentry.mydomain.com``
is mapped to your docker host) and login with default credentials
(username ``admin`` and password ``admin``) and create your first team and project.

Your sqlite database file and web workers logs are available in ``/tmp/sentry`` directory.

## Contributing

Try not to fork this repo just to create your own Docker image with some
minor tweak. Please open an [issue on GitHub](https://github.com/slafs/sentry-docker/issues)
and maybe we can include your use case directly within this image :).

You can even write a test case for your feature ;). See
[CONTRIBUTING.md](https://github.com/slafs/sentry-docker/blob/master/CONTRIBUTING.md).

Also feel free to give [feedback and comments](https://github.com/slafs/sentry-docker/issues)
about this image in general.

## Advanced usage

Copy the file with environment variables ``environment.example`` e.g ``cp environment.example environment``
and after tweaking some values run sentry like this

```
docker run -d --name=sentry --volume=/tmp/sentry:/data -p 80:9000 --env-file=environment slafs/sentry
```

``ENTRYPOINT`` for this image is a little wrapper script around default ``sentry`` executable.
Default ``CMD`` is ``start`` which runs ``upgrade`` command (initializes and upgrades the database)
and creates a default administrator (superuser) and then runs a http service (as in ``sentry --config=... start``).

All commands and args are passed to the ``sentry`` executable. This means

```
docker run [docker options] slafs/sentry --help
```

refers to running:

```
sentry --config=/conf/sentry.conf.py --help
```

inside the container.

### Admin user

You can specify a username, password and email address to create an initial sentry administrator.

Add those variables to your environment file

```
SENTRY_ADMIN_USERNAME=slafs
SENTRY_ADMIN_PASSWORD=mysecretpass
SENTRY_ADMIN_EMAIL=slafs@foo.com
```

See [django docs](https://docs.djangoproject.com/en/1.5/ref/django-admin/#createsuperuser) for details about ``createsuperuser`` command.

The default administrator username is ``admin`` with password ``admin`` (and ``root@localhost`` as email adress).

### Postgres

It is recommended to run your sentry instance with PostgreSQL database.
To link your sentry instance with a postgres container you can do it like this:

0. Pull an official PostgreSQL image: ``docker pull postgres`` (if you haven't already).
1. Run postgres container (from the official Docker image): ``docker run -d --name=postgres_container postgres``.
2. Add ``DATABASE_URL=postgres://postgres:@postgresdb/postgres`` to environment file.
3. Run sentry with linked postgres container: ```
docker run -d --name=sentry --volume=/tmp/sentry:/data -p 80:9000 --env-file=environment --link=postgres_container:postgresdb slafs/sentry```

Notice that an alias for your linked postgres container (``postgresdb``) is the same as a postgres host in ``DATABASE_URL`` variable.

``DATABASE_URL`` is a value that is parsed by an external app called [dj-database-url](https://github.com/kennethreitz/dj-database-url).


### Redis

Redis container can be used to improve sentry performance in number of ways. It can be used as a:

* cache backend,
* sentry buffers,
* celery broker.

You can link your sentry instance with a redis container like this:

0. Pull an official Redis image: ``docker pull redis`` (if you haven't already).
1. Run redis container (from the official Docker image): ``docker run -d --name=redis_container redis``
2. ```
docker run -d --name=sentry --volume=/tmp/sentry:/data -p 80:9000 --env-file=environment --link=postgres_container:postgresdb --link=redis_container:redis slafs/sentry```

If you want to use a different container alias for redis you should add ```SENTRY_REDIS_HOST=your_redis_alias``` to environment file.

#### Cache with redis

To use a redis cache backend add ``CACHE_URL=hiredis://redis:6379/2``
(make sure you won't have a trailing `/` in this setting)
to environment file (where ``redis`` is the alias of your linked redis container).
See [django-cache-url](https://github.com/ghickman/django-cache-url) docs for available formats.

#### Sentry buffers with redis

To use [sentry update buffers](https://docs.getsentry.com/on-premise/server/buffer/)
with redis you must add ``SENTRY_USE_REDIS_BUFFERS=True`` to environment file.

If you have many redis containers/hosts you can set a list of those hosts
in ``SENTRY_REDIS_BUFFERS`` variable so they can be used by sentry.
Like this: ``SENTRY_REDIS_BUFFERS=redis1:6380,redis2:6381``.

See [sentry docs](https://docs.getsentry.com/on-premise/server/buffer/#redis) for details about redis buffer.

#### Celery with redis (and postgres)

To use Celery in sentry you must add ``CELERY_ALWAYS_EAGER=False`` to your environment file and run a celery worker like this:

```
docker run -d --name=sentry_celery_worker --link=redis_container:redis --link=postgres_container:postgresdb --volume=/tmp/sentry:/data --env-file=environment slafs/sentry celery worker -B
```

You can also set a different ``BROKER_URL`` via environment file by adding this:
``SENTRY_BROKER_URL=redis://otherredishost:6379/1``

You can run as many celery worker containers as you want but remember that only one of them should be run with ``-B`` option.

If you're using default values for `CELERY_RESULT_SERIALIZER`, `CELERY_TASK_SERIALIZER`
and `CELERY_ACCEPT_CONTENT` (default values support `pickle`)
you have to set `C_FORCE_ROOT` env var (to say `1`)
to be able to run celery as a `root` user.
You can avoid this by setting those three to `json`.
Refer to [Celery's serializer docs](docs.celeryproject.org/en/latest/userguide/security.html#serializers)
for more info about security implications of using `pickle` as a task serialization format.

#### Time-series storage with redis

To have [time-series data](https://docs.getsentry.com/on-premise/server/tsdb/) you must add
``SENTRY_USE_REDIS_TSDB=True`` to the environment file.

By default Redis time-series storage will use the Redis connection
defined by ``SENTRY_REDIS_HOST`` and ``SENTRY_REDIS_PORT``. Similarly
to configuring buffers, you can set ``SENTRY_REDIS_TSDBS`` to a list
of Redis servers: ``SENTRY_REDIS_TSDBS=redis1:6379,redis2:6380``

### Email

You can configure all [email settings](https://docs.getsentry.com/on-premise/server/config/#smtp-server)
by environment variables with ``SENTRY_`` prefix.
You have to also change an email backend and set it
to ``SENTRY_EMAIL_BACKEND=django.core.mail.backends.smtp.EmailBackend`` or something similar.

### LDAP
With this image You should be able to easily configure LDAP authentication for your Sentry instance.
To enable it add ``SENTRY_USE_LDAP=True`` to your ``environment`` file.
Then set the needed options by adding env variables with ``LDAP_``
prefix (see the table below). LDAP authentication backend is provided by
[django-auth-ldap](https://pythonhosted.org/django-auth-ldap/).

### REMOTE_USER
To enable authentication via REMOTE_USER, add ``SENTRY_USE_REMOTE_USER=True`` to your ``environment`` file.
See the ``AUTH_REMOTE_USER_*`` env variables below for further configuration.

## Available environment variables

Refer to [sentry documentation](https://docs.getsentry.com/on-premise/server/config/),
[django documentation](https://docs.djangoproject.com/en/1.6/ref/settings/),
[celery documentation](http://docs.celeryproject.org/en/latest/)
and [django-auth-ldap documentation](https://pythonhosted.org/django-auth-ldap/reference.html)
for the meaning of each setting.

Environment variable name       | Django/Sentry setting                         | Type | Default value                                         | Description
--------------------------------|-----------------------------------------------|------|-------------------------------------------------------|------------------------------------------------------------------------
SECRET_KEY                      | SECRET_KEY                                    |      | **REQUIRED!**                                         | set this to something random
SENTRY_URL_PREFIX               | SENTRY_URL_PREFIX                             |      | **REQUIRED!**                                         | no trailing slash!
DATABASE_URL                    | DATABASES                                     |      | sqlite:////data/sentry.db                             |
CACHE_URL                       | CACHES                                        |      | locmem://                                             |
CELERY_ALWAYS_EAGER             | CELERY_ALWAYS_EAGER                           | bool | True                                                  |
SENTRY_BROKER_URL               | BROKER_URL                                    |      | ``redis://<SENTRY_REDIS_HOST>:<SENTRY_REDIS_PORT>/1`` |
CELERY_RESULT_SERIALIZER        | CELERY_RESULT_SERIALIZER                      |      | ``pickle``                                            | You may want change it to ``json``. See http://docs.celeryproject.org/en/stable/configuration.html#celery-result-serializer
CELERY_TASK_SERIALIZER          | CELERY_TASK_SERIALIZER                        |      | ``pickle``                                            | You may want change it to ``json``. See http://docs.celeryproject.org/en/stable/configuration.html#celery-task-serializer
CELERY_ACCEPT_CONTENT           | CELERY_ACCEPT_CONTENT                         | list | ``pickle,json``                                       | Comma separated values. You may want change it to ``json``. See http://docs.celeryproject.org/en/stable/configuration.html#celery-accept-content
SENTRY_REDIS_HOST               |                                               |      | redis                                                 |
SENTRY_REDIS_PORT               |                                               | int  | 6379                                                  |
SENTRY_WEB_HOST                 | SENTRY_WEB_HOST                               |      | 0.0.0.0                                               |
SENTRY_WEB_PORT                 | SENTRY_WEB_PORT                               | int  | 9000                                                  |
SENTRY_WORKERS                  | SENTRY_WEB_OPTIONS['workers']                 | int  | 3                                                     | the number of sentry web workers
SENTRY_USE_REDIS_BUFFERS        |                                               | bool | False                                                 |
SENTRY_REDIS_BUFFERS            | SENTRY_REDIS_OPTIONS['hosts']*                | list | ``<SENTRY_REDIS_HOST>:<SENTRY_REDIS_PORT>``           | comma separated list of redis hosts (``host1:port1,host2:port2,...``)
SENTRY_USE_REDIS_TSDB           |                                               | bool | False                                                 |
SENTRY_REDIS_TSDBS              | SENTRY_TSDB_OPTIONS['hosts']*                 | list | ``<SENTRY_REDIS_HOST>:<SENTRY_REDIS_PORT>``           | comma separated list of redis hosts (``host1:port1,host2:port2,...``)
SENTRY_EMAIL_BACKEND            | EMAIL_BACKEND                                 |      | django.core.mail.backends.console.EmailBackend        |
SENTRY_EMAIL_HOST               | EMAIL_HOST                                    |      | localhost                                             |
SENTRY_EMAIL_HOST_PASSWORD      | EMAIL_HOST_PASSWORD                           |      | ''                                                    |
SENTRY_EMAIL_HOST_USER          | EMAIL_HOST_USER                               |      | ''                                                    |
SENTRY_EMAIL_PORT               | EMAIL_PORT                                    | int  | 25                                                    |
SENTRY_EMAIL_USE_TLS            | EMAIL_USE_TLS                                 | bool | False                                                 |
SENTRY_SERVER_EMAIL             | SERVER_EMAIL                                  |      | root@localhost                                        |
SENTRY_ALLOW_REGISTRATION       | SENTRY_ALLOW_REGISTRATION                     | bool | False                                                 |
SENTRY_ADMIN_USERNAME           |                                               |      | admin                                                 | username for Sentry's superuser
SENTRY_ADMIN_PASSWORD           |                                               |      | admin                                                 | password for Sentry's superuser
SENTRY_ADMIN_EMAIL              | SENTRY_ADMIN_EMAIL                            |      | root@localhost                                        | email address for Sentry's superuser and a setting as of Sentry 7.3
SENTRY_DATA_DIR                 |                                               |      | ``/data``                                             | custom location for logs and sqlite database
TWITTER_CONSUMER_KEY            | TWITTER_CONSUMER_KEY                          |      | ''                                                    |
TWITTER_CONSUMER_SECRET         | TWITTER_CONSUMER_SECRET                       |      | ''                                                    |
FACEBOOK_APP_ID                 | FACEBOOK_APP_ID                               |      | ''                                                    |
FACEBOOK_API_SECRET             | FACEBOOK_API_SECRET                           |      | ''                                                    |
GOOGLE_OAUTH2_CLIENT_ID         | GOOGLE_OAUTH2_CLIENT_ID                       |      | ''                                                    |
GOOGLE_OAUTH2_CLIENT_SECRET     | GOOGLE_OAUTH2_CLIENT_SECRET                   |      | ''                                                    |
GITHUB_APP_ID                   | GITHUB_APP_ID                                 |      | ''                                                    |
GITHUB_API_SECRET               | GITHUB_API_SECRET                             |      | ''                                                    |
TRELLO_API_KEY                  | TRELLO_API_KEY                                |      | ''                                                    |
TRELLO_API_SECRET               | TRELLO_API_SECRET                             |      | ''                                                    |
BITBUCKET_CONSUMER_KEY          | BITBUCKET_CONSUMER_KEY                        |      | ''                                                    |
BITBUCKET_CONSUMER_SECRET       | BITBUCKET_CONSUMER_SECRET                     |      | ''                                                    |
SENTRY_USE_LDAP                 |                                               | bool | False                                                 | if set to ``False`` all other LDAP settings are discarded
LDAP_SERVER                     | AUTH_LDAP_SERVER_URI                          |      | ``ldap://localhost``                                  |
LDAP_BIND_DN                    | AUTH_LDAP_BIND_DN                             |      | ''                                                    |
LDAP_BIND_PASSWORD              | AUTH_LDAP_BIND_PASSWORD                       |      | ''                                                    |
LDAP_USER_DN                    | AUTH_LDAP_USER_SEARCH*                        |      | **REQUIRED!** if you want to use LDAP auth            | first argument of LDAPSearch (base_dn) when searching for users
LDAP_USER_FILTER                | AUTH_LDAP_USER_SEARCH*                        |      | ``(&(objectClass=inetOrgPerson)(cn=%(user)s))``       | third argument of LDAPSearch (filterstr) when searching for users
LDAP_GROUP_DN                   | AUTH_LDAP_GROUP_SEARCH*                       |      | ''                                                    | first argument of LDAPSearch (base_dn) when searching for groups
LDAP_GROUP_FILTER               | AUTH_LDAP_GROUP_SEARCH*                       |      | ``(objectClass=groupOfUniqueNames)``                  | third argument of LDAPSearch (filterstr) when searching for groups
LDAP_GROUP_TYPE                 | AUTH_LDAP_GROUP_TYPE*                         |      | ''                                                    | if set to 'groupOfUniqueNames' then ``AUTH_LDAP_GROUP_TYPE = GroupOfUniqueNamesType()``, if set to 'posixGroup' then ``AUTH_LDAP_GROUP_TYPE = PosixGroupType()``.
LDAP_REQUIRE_GROUP              | AUTH_LDAP_REQUIRE_GROUP                       |      | None                                                  |
LDAP_DENY_GROUP                 | AUTH_LDAP_DENY_GROUP                          |      | None                                                  |
LDAP_MAP_FIRST_NAME             | AUTH_LDAP_USER_ATTR_MAP['first_name']         |      | ``givenName``                                         |
LDAP_MAP_LAST_NAME              | AUTH_LDAP_USER_ATTR_MAP['last_name']          |      | ``sn``                                                |
LDAP_MAP_MAIL                   | AUTH_LDAP_USER_ATTR_MAP['email']              |      | ``mail``                                              |

CHANGELOG.md
# Changelog

## 2016-03-07

- Upgraded to Sentry ``8.0``

## 2015-12-18

- added `CELERY_RESULT_SERIALIZER`, `CELERY_TASK_SERIALIZER`
  and `CELERY_ACCEPT_CONTENT` configs to avoid `C_FORCE_ROOT` env var
  (default values support `pickle` for backwards compatibility).

## 2015-11-16

- updated ``Sentry`` to 7.7.4
- merged forgotten changes to dev

## 2015-11-11

- updated ``Sentry`` to 7.7.1
- updated ``django-auth-ldap`` to 1.2.7
- updated ``django-redis`` to 4.3.0
- updated ``python-decouple`` to 3.0
- fixed building dev version (8.0.0.dev0) (credit goes to: @yurtaev)
- creating superuser via python script

## 2015-08-29

- updated ``Sentry`` to 7.7.0 and dropped ``django-json`` fix

## 2015-08-26

- updated ``Sentry`` to 7.6.2

## 2015-08-24

- add `.dockerignore` file
- updated ``django-cache-url`` to 1.0.0 (watch out for trailing `/`)
- updated ``django-redis`` to 4.2.0

## 2015-08-20

- Added `nestedGroupOfNames` `LDAP_GROUP_TYPE`
- Make use of `LDAP_GROUP_STAFF`

## 2015-08-19

- fix migration issue (https://github.com/getsentry/sentry/issues/1648)
- allow for a simple database accessibility check (via ``SENTRY_DOCKER_DO_DB_CHECK`` env var)
- updated ``Sentry`` to 7.5.6

## 2015-06-25

- remove ``nydus`` as an explicit requirement (see #28)

## 2015-06-22

- updated ``Sentry`` to 7.5.4

## 2015-06-08

- revert django-redis to 3.8.4 due to #26

## 2015-05-20

- updated ``requirements.txt`` (``django-redis``, ``hiredis`` and ``django-auth-ldap`` )
- added support for posixGroup in LDAP auth backend (credit goes to: @grundleborg)

## 2015-03-26

- updated ``Sentry`` to 7.4.3

## 2015-03-17

- Add support for REMOTE_USER authentication (credit goes to: @abesto)

## 2015-03-16

- Add support for SENTRY_PUBLIC (credit goes to: @abesto)

## 2015-03-12

- Add support for configuring time-series storage (credit goes to:
  @abesto)

## 2015-03-07

- Properly pass SIGTERM to sentry process (credit goes to: @lorenz)
- new Sentry 7.4.X
- move from fig to docker-compose while testing
- when ``SENTRY_USE_REDIS_BUFFERS`` is used then ``sentry.cache.redis.RedisCache`` is configured as a ``SENTRY_CACHE``

## 2015-02-23

- updated ``requirements.txt`` (``django-redis``)
- new Sentry 7.3.X

## 2015-02-12

- updated ``requirements.txt`` (``django-auth-ldap`` and ``hiredis``)
- new Sentry 7.2.X

## 2015-01-27

- new tag/branch ``7.1``
- updated ``requirements.txt`` (``django-redis``)

## 2015-01-08

- updated ``requirements.txt`` (``django-redis``)

## 2015-01-04

- updated ``requirements.txt`` (``django-auth-ldap``)
- new Sentry 7.0.X and several new build tags on docker hub

## 2014-12-07

- updated ``django-redis`` to 3.8.0
- provide builds from sentry's master github branch

## 2014-12-01

- added ``SENTRY_ALLOW_ORIGIN`` - thanks to @josh-devops-center

## 2014-11-30

- FIX: determine platform when creating a project (via ``SENTRY_INITIAL_PLATFORM``) - thanks to @josh-devops-center

## 2014-11-20

- added ``SENTRY_INTIAL_DOMAINS`` for allowed domains (see #7)
- allow setting ``SENTRY_SECURE_PROXY_SSL_HEADER`` and ``SENTRY_USE_X_FORWARDED_HOST`` (see #6)

## 2014-11-05

- update django-redis requirement
- add some cleanup at the end of build process
- add new env vars for initial team, project and project key (``SENTRY_INITIAL_*``)
- add new command ``prepare`` for the wrapper script to just prepare the datbase
  and other stuff without running the http sevice.

## 2014-10-09

- change the name of the config script (this is a possible *breaking change*
  but now you can easily extend this image and configuration)
- make it possible to "test" this image

## 2014-10-06

- updated ``requirements.txt`` (``django-auth-ldap`` and ``hiredis``)
- added some contribution guidelines

## 2014-09-09

- added new setting ``SENTRY_ALLOW_REGISTRATION`` (credit goes to: @lukas-hetzenecker)
- added LDAP backend support (``SENTRY_USE_LDAP`` and ``LDAP_*`` settings)
  (credit goes to: @lukas-hetzenecker)



sentry_docker_conf.py
# This file is just Python, with a touch of Django which means you
# you can inherit and tweak settings to your hearts content.
from sentry.conf.server import *  # noqa
import os.path

from decouple import config
import dj_database_url
import django_cache_url
import functools

CONF_ROOT = os.path.dirname(__file__)
DATA_DIR = config('SENTRY_DATA_DIR', default='/data')
DEFAULT_SQLITE_DB_PATH = os.path.join(DATA_DIR, 'sentry.db')

REDIS_HOST = config('SENTRY_REDIS_HOST', default='redis')
REDIS_PORT = config('SENTRY_REDIS_PORT', default=6379, cast=int)

DATABASES = {
    'default': dj_database_url.config(default='sqlite:///{0}'.format(DEFAULT_SQLITE_DB_PATH))
}

if 'postgres' in DATABASES['default']['ENGINE']:
    DATABASES['default']['OPTIONS'] = {
        'autocommit': True,
    }

CACHES = {'default': django_cache_url.config() }
SENTRY_CACHE = 'sentry.cache.django.DjangoCache'

SENTRY_PUBLIC = config('SENTRY_PUBLIC', default=False, cast=bool)

def nydus_config(from_env_var):
    """
    Generate a Nydus Redis configuration from an ENV variable of the form "server:port,server:port..."
    Default to REDIS_HOST:REDIS_PORT if the ENV variable is not provided.
    """
    redis_servers_cast = lambda x: list(r.split(':') for r in x.split(','))
    servers = config(from_env_var, default='{0}:{1}'.format(REDIS_HOST, REDIS_PORT), cast=redis_servers_cast)
    _redis_hosts = {}

    for r_index, r_host_pair in enumerate(servers):
        _redis_hosts[r_index] = {'host': r_host_pair[0], 'port': int(r_host_pair[1])}

    return {
        'hosts': _redis_hosts
    }

############################
# General Sentry options ##
############################
SENTRY_OPTIONS = {
    'system.url-prefix': config('SENTRY_URL_PREFIX'),
    'system.admin-email': config('SENTRY_ADMIN_EMAIL', default='root@localhost'),
}

###########
# Queue ##
###########

# See http://sentry.readthedocs.org/en/latest/queue/index.html for more
# information on configuring your queue broker and workers. Sentry relies
# on a Python framework called Celery to manage queues.

# You can enable queueing of jobs by turning off the always eager setting:
CELERY_ALWAYS_EAGER = config('CELERY_ALWAYS_EAGER', default=True, cast=bool)
DEFAULT_BROKER_URL = 'redis://{0}:{1}/1'.format(REDIS_HOST, REDIS_PORT)

BROKER_URL = config('SENTRY_BROKER_URL', default=DEFAULT_BROKER_URL)

CELERY_RESULT_SERIALIZER = config('CELERY_RESULT_SERIALIZER', default='pickle')
CELERY_TASK_SERIALIZER = config('CELERY_TASK_SERIALIZER', default='pickle')
CELERY_ACCEPT_CONTENT = config('CELERY_ACCEPT_CONTENT', default='pickle,json', cast=lambda x: x.split(','))

####################
# Update Buffers ##
####################

# Buffers (combined with queueing) act as an intermediate layer between the
# database and the storage API. They will greatly improve efficiency on large
# numbers of the same events being sent to the API in a short amount of time.
# (read: if you send any kind of real data to Sentry, you should enable buffers)

# You'll need to install the required dependencies for Redis buffers:
#   pip install redis hiredis nydus
#

SENTRY_USE_REDIS_BUFFERS = config('SENTRY_USE_REDIS_BUFFERS', default=False, cast=bool)

if SENTRY_USE_REDIS_BUFFERS:
    SENTRY_BUFFER = 'sentry.buffer.redis.RedisBuffer'
    SENTRY_REDIS_OPTIONS = nydus_config('SENTRY_REDIS_BUFFERS')
    SENTRY_CACHE = 'sentry.cache.redis.RedisCache'

#######################
# Time-series Storage #
#######################
# Sentry provides a service to store time-series data. Primarily this
# is used to display aggregate information for events and projects, as
# well as calculating (in real-time) the rates of events.

SENTRY_USE_REDIS_TSDB = config('SENTRY_USE_REDIS_TSDB', default=False, cast=bool)

if SENTRY_USE_REDIS_TSDB:
    SENTRY_TSDB = 'sentry.tsdb.redis.RedisTSDB'
    SENTRY_TSDB_OPTIONS = nydus_config('SENTRY_REDIS_TSDBS')

################
# Web Server ##
################

# If you're using a reverse proxy, you should enable the X-Forwarded-Proto
# and X-Forwarded-Host headers, and uncomment the following settings
SECURE_PROXY_SSL_HEADER = config('SENTRY_SECURE_PROXY_SSL_HEADER', default=None, cast=lambda x: tuple(x.split(',')) if x else None)
USE_X_FORWARDED_HOST = config('SENTRY_USE_X_FORWARDED_HOST', default=False, cast=bool)

SENTRY_WEB_HOST = config('SENTRY_WEB_HOST', default='0.0.0.0')
SENTRY_WEB_PORT = config('SENTRY_WEB_PORT', default=9000, cast=int)
SENTRY_WEB_OPTIONS = {
    'workers': config('SENTRY_WORKERS', default=3, cast=int),  # the number of gunicorn workers
    'limit_request_line': 0,  # required for raven-js
    'secure_scheme_headers': {'X-FORWARDED-PROTO': 'https'},
    'errorlog' : os.path.join(DATA_DIR, 'gunicorn_error.log'),
    'accesslog' : os.path.join(DATA_DIR, 'gunicorn_access.log'),
}

# allows JavaScript clients to submit cross-domain error reports. Useful for local development
SENTRY_ALLOW_ORIGIN = config('SENTRY_ALLOW_ORIGIN', default=None)

#################
# Mail Server ##
#################

# For more information check Django's documentation:
#  https://docs.djangoproject.com/en/1.3/topics/email/?from=olddocs#e-mail-backends

EMAIL_BACKEND = config('SENTRY_EMAIL_BACKEND', default='django.core.mail.backends.console.EmailBackend')

EMAIL_HOST = config('SENTRY_EMAIL_HOST', default='localhost')
EMAIL_HOST_PASSWORD = config('SENTRY_EMAIL_HOST_PASSWORD', default='')
EMAIL_HOST_USER = config('SENTRY_EMAIL_HOST_USER', default='')
EMAIL_PORT = config('SENTRY_EMAIL_PORT', default=25, cast=int)
EMAIL_USE_TLS = config('SENTRY_EMAIL_USE_TLS', default=False, cast=bool)

# The email address to send on behalf of
SERVER_EMAIL = config('SENTRY_SERVER_EMAIL', default='root@localhost')

###########
# etc. ##
###########

SENTRY_FEATURES['auth:register'] = config('SENTRY_ALLOW_REGISTRATION', default=False, cast=bool)

# If this file ever becomes compromised, it's important to regenerate your SECRET_KEY
# Changing this value will result in all current sessions being invalidated
SECRET_KEY = config('SECRET_KEY')

# http://twitter.com/apps/new
# It's important that input a callback URL, even if its useless. We have no idea why, consult Twitter.
TWITTER_CONSUMER_KEY = config('TWITTER_CONSUMER_KEY', default='')
TWITTER_CONSUMER_SECRET = config('TWITTER_CONSUMER_SECRET', default='')

# http://developers.facebook.com/setup/
FACEBOOK_APP_ID = config('FACEBOOK_APP_ID', default='')
FACEBOOK_API_SECRET = config('FACEBOOK_API_SECRET', default='')

# http://code.google.com/apis/accounts/docs/OAuth2.html#Registering
GOOGLE_OAUTH2_CLIENT_ID = config('GOOGLE_OAUTH2_CLIENT_ID', default='')
GOOGLE_OAUTH2_CLIENT_SECRET = config('GOOGLE_OAUTH2_CLIENT_SECRET', default='')

# https://github.com/settings/applications/new
GITHUB_APP_ID = config('GITHUB_APP_ID', default='')
GITHUB_API_SECRET = config('GITHUB_API_SECRET', default='')

# https://trello.com/1/appKey/generate
TRELLO_API_KEY = config('TRELLO_API_KEY', default='')
TRELLO_API_SECRET = config('TRELLO_API_SECRET', default='')

# https://confluence.atlassian.com/display/BITBUCKET/OAuth+Consumers
BITBUCKET_CONSUMER_KEY = config('BITBUCKET_CONSUMER_KEY', default='')
BITBUCKET_CONSUMER_SECRET = config('BITBUCKET_CONSUMER_SECRET', default='')

# custom settings
ALLOWED_HOSTS = ['*']
INTERNAL_IPS = config('INTERNAL_IPS', default='') # TODO: Milan check of dit werkt
LOGGING['disable_existing_loggers'] = False

SENTRY_BEACON = config('SENTRY_BEACON', default=True, cast=bool)

####################
# LDAP settings ##
####################

SENTRY_USE_LDAP = config('SENTRY_USE_LDAP', default=False, cast=bool)

if SENTRY_USE_LDAP:
    import ldap
    from django_auth_ldap.config import LDAPSearch, GroupOfUniqueNamesType, PosixGroupType, NestedGroupOfNamesType

    AUTH_LDAP_SERVER_URI = config('LDAP_SERVER', default='ldap://localhost')

    AUTH_LDAP_BIND_DN = config('LDAP_BIND_DN', default='')
    AUTH_LDAP_BIND_PASSWORD = config('LDAP_BIND_PASSWORD', default='')

    AUTH_LDAP_USER_SEARCH = LDAPSearch(
        config('LDAP_USER_DN'),
        ldap.SCOPE_SUBTREE,
        config('LDAP_USER_FILTER', default='(&(objectClass=inetOrgPerson)(cn=%(user)s))')
    )

    AUTH_LDAP_GROUP_SEARCH = LDAPSearch(
        config('LDAP_GROUP_DN', default=''),
        ldap.SCOPE_SUBTREE,
        config('LDAP_GROUP_FILTER', default='(objectClass=groupOfUniqueNames)')
    )

    if config('LDAP_GROUP_TYPE', default='') == 'groupOfUniqueNames':
        AUTH_LDAP_GROUP_TYPE = GroupOfUniqueNamesType()
    elif config('LDAP_GROUP_TYPE', default='') == 'posixGroup':
        AUTH_LDAP_GROUP_TYPE = PosixGroupType()
    elif config('LDAP_GROUP_TYPE', default='') == 'nestedGroupOfNames':
        AUTH_LDAP_GROUP_TYPE = NestedGroupOfNamesType()

    AUTH_LDAP_REQUIRE_GROUP = config('LDAP_REQUIRE_GROUP', default=None)
    AUTH_LDAP_DENY_GROUP = config('LDAP_DENY_GROUP', default=None)

    AUTH_LDAP_USER_ATTR_MAP = {
        'first_name': config('LDAP_MAP_FIRST_NAME', default='givenName'),
        'last_name': config('LDAP_MAP_LAST_NAME', default='sn'),
        'email': config('LDAP_MAP_MAIL', default='mail')
    }

    ldap_is_active    = config('LDAP_GROUP_ACTIVE', default='')
    ldap_is_superuser = config('LDAP_GROUP_SUPERUSER', default='')
    ldap_is_staff     = config('LDAP_GROUP_STAFF', default='')

    if ldap_is_active or ldap_is_superuser or ldap_is_staff:
        AUTH_LDAP_USER_FLAGS_BY_GROUP = {
            'is_active': ldap_is_active,
            'is_superuser': ldap_is_superuser,
            'is_staff': ldap_is_staff,
        }

    AUTH_LDAP_FIND_GROUP_PERMS = config('LDAP_FIND_GROUP_PERMS', default=False, cast=bool)

    # Cache group memberships for an hour to minimize LDAP traffic
    AUTH_LDAP_CACHE_GROUPS = config('LDAP_CACHE_GROUPS', default=True, cast=bool)
    AUTH_LDAP_GROUP_CACHE_TIMEOUT = config('LDAP_GROUP_CACHE_TIMEOUT', default=3600, cast=int)

    AUTHENTICATION_BACKENDS = AUTHENTICATION_BACKENDS + (
        'django_auth_ldap.backend.LDAPBackend',
    )

    # setup logging for django_auth_ldap
    import logging
    logger = logging.getLogger('django_auth_ldap')
    logger.addHandler(logging.StreamHandler())
    ldap_loglevel = getattr(logging, config('LDAP_LOGLEVEL', default='DEBUG'))
    logger.setLevel(ldap_loglevel)

##############################
# REMOTE_USER authentication #
##############################

SENTRY_USE_REMOTE_USER = config('SENTRY_USE_REMOTE_USER', default=False, cast=bool)

if SENTRY_USE_REMOTE_USER:
    AUTHENTICATION_BACKENDS += ('django.contrib.auth.backends.RemoteUserBackend',)

    AUTH_REMOTE_USER_HEADER = config('AUTH_REMOTE_USER_HEADER', default=None)
    if AUTH_REMOTE_USER_HEADER:
        # The lazy hack is required because importing RemoteUserMiddleware at load time leads to a circular import
        # The name is upper camel case because that's the only way to expose values from this config file
        def build_LAZY_CUSTOM_REMOTE_USER_MIDDLEWARE(header, *args, **kwargs):
            from django.contrib.auth.middleware import RemoteUserMiddleware
            class CustomRemoteUserMiddleware(RemoteUserMiddleware):
               pass
            CustomRemoteUserMiddleware.header = header
            return CustomRemoteUserMiddleware(*args, **kwargs)
        LAZY_CUSTOM_REMOTE_USER_MIDDLEWARE = functools.partial(build_LAZY_CUSTOM_REMOTE_USER_MIDDLEWARE, AUTH_REMOTE_USER_HEADER)
        MIDDLEWARE_CLASSES += ('sentry_config.LAZY_CUSTOM_REMOTE_USER_MIDDLEWARE',)
    else:
        MIDDLEWARE_CLASSES += ('django.contrib.auth.middleware.RemoteUserMiddleware',)
